app/tests/heartbeat_controller_desktop_test.cc

```diff
@@ -336,5 +344,291 @@ TEST_F(HeartbeatControllerDesktopTest,
+          agent: "agent/1",
+          dates: ["2000-01-23", "2000-01-24"]
+        }
+      ],
+      "version":"2"
+    })json"));
+
+  // Storage should still have last_logged_date, but the heartbeats should no
+  // longer be stored.
+  LoggedHeartbeats read_heartbeats;
+  bool read_ok = storage_.ReadTo(read_heartbeats);
+  ASSERT_TRUE(read_ok);
+  EXPECT_EQ(read_heartbeats.last_logged_date, day2);
+  EXPECT_EQ(read_heartbeats.heartbeats.size(), 0);
+}
+
+TEST_F(HeartbeatControllerDesktopTest, GetTodaysHeartbeatThenGetAllHeartbeats) {
+  app_common::RegisterLibrariesFromUserAgent(kDefaultUserAgent);
+  std::string day1 = "2000-01-23";
+  std::string day2 = "2000-01-24";
+  // Date provider will be called twice for, Log then twice for Get
+  EXPECT_CALL(mock_date_provider_, GetDate())
+      .Times(4)
+      .WillOnce(Return(day1))
+      .WillRepeatedly(Return(day2));
+
+  controller_.LogHeartbeat();
+  controller_.LogHeartbeat();
+  // GetAndResetStoredHeartbeats is done synchronously, so there is no need to
+  // wait.
+  std::string payload = controller_.DecodeAndDecompress(
+      controller_.GetAndResetTodaysStoredHeartbeats());
+
+  EXPECT_THAT(payload, EqualsJson(R"json({
+      "heartbeats": [
+        {
+          agent: "agent/1",
+          dates: ["2000-01-24"]
+        }
+      ],
+      "version":"2"
+    })json"));
+
+  // The heartbeat for the previous day should still be stored (01-23).
+  std::string payload2 = controller_.DecodeAndDecompress(
+      controller_.GetAndResetStoredHeartbeats());
+
+  EXPECT_THAT(payload2, EqualsJson(R"json({
+      "heartbeats": [
+        {
+          agent: "agent/1",
+          dates: ["2000-01-23"]
+        }
+      ],
+      "version":"2"
+    })json"));
+}
+
+TEST_F(HeartbeatControllerDesktopTest, GetHeartbeatPayloadMultipleTimes) {
+  // Override the min time between fetches to allow multiple fetches
+  double time_between_fetches_original_val = g_min_time_between_fetches_sec;
+  g_min_time_between_fetches_sec = 0.0;
+
+  app_common::RegisterLibrariesFromUserAgent(kDefaultUserAgent);
+  std::string today = "2000-01-23";
+  // Date provider will be called for both Log and Get
+  EXPECT_CALL(mock_date_provider_, GetDate())
+      .Times(3)
+      .WillRepeatedly(Return(today));
+
+  controller_.LogHeartbeat();
+  // GetAndResetStoredHeartbeats is done synchronously, so there is no need to
+  // wait.
+  std::string first_payload = controller_.DecodeAndDecompress(
+      controller_.GetAndResetStoredHeartbeats());
+  EXPECT_THAT(first_payload, EqualsJson(R"json({
+      "heartbeats": [
+        {
+          agent: "agent/1",
+          dates: ["2000-01-23"]
+        }
+      ],
+      "version":"2"
+    })json"));
+
+  std::string second_payload = controller_.DecodeAndDecompress(
+      controller_.GetAndResetStoredHeartbeats());
+  EXPECT_EQ(second_payload, "");
+  // Reset the time between fetches to original value
+  g_min_time_between_fetches_sec = time_between_fetches_original_val;
```

Comment:
You can put the following lines in `HeartbeatControllerDesktopTest::SetUp()`

```
  void SetUp() override {
    // Reset the time between fetches to original value
    g_min_time_between_fetches_sec = time_between_fetches_original_val;
  }
```

Reply:
as mentioned in chat, I put this in the TearDown instead so that it happens after the test instead of before.